File: euler_096.py
from time import time
import numpy as np

def possible(grid, val, r, c):

    # Grid cell must be empty
    if grid[r, c] != 0:
        return False

    # Row or column cannot contain val already
    if np.count_nonzero(grid[r, :] == val) + np.count_nonzero(grid[:, c] == val) > 0:
        return False

    # Same 3x3 grid cannot contain val already
    if np.count_nonzero(grid[3 * (r // 3) : 3 * (r // 3) + 3, 3 * (c // 3) : 3 * (c // 3) + 3] == val) > 0:
        return False

    return True

def possibles(grid, r, c):
    return [val for val in range(1, 10) if possible(grid, val, r, c)]

def solve(n):

    global grids

    rows, cols = np.where(grids[n] == 0)

    for r in rows:
        for c in cols:
            for val in possibles(grids[n], r, c):
                grids[n][r, c] = val
                if solve(n):
                    return True
                grids[n][r, c] = 0
            return

    return True



start = time()

# Importing grids from file
# ---------------------------------------
with open("p096_sudoku.txt", "r") as f:
    raw = f.readlines()
    raw_stripped = []
    for i in range(len(raw)):
        if i % 10 > 0:
            raw_stripped.append(raw[i].replace("\n", ""))
    grids = []
    for gn in range(50):
        grid = []
        for i in range(9):
            grid.append([int(n) for n in raw_stripped[9 * gn + i]])
        grids.append(np.array(grid))
# ---------------------------------------

# Filling out the easy cells (1 possibility)
for i in range(3):
    for n in range(len(grids)):
        grid = grids[n]
        for r in range(9):
            for c in range(9):
                p_list = possibles(grid, r, c)
                if len(p_list) == 1:
                    grid[r, c] = p_list[0]

total = 0

for n in range(len(grids)):
    solve(n)
    total += grids[n][0, 0] * 100 + grids[n][0, 1] * 10 + grids[n][0, 2]

print("Sum of corners: ", total)

stop = time()
print("Time: " + str(stop - start))

File: euler_096.js
const start = Date.now();

import { readFileSync } from 'fs';

function syncReadFile(filename) {
    const contents = readFileSync(filename, 'utf-8');

    const arr = contents.split("\n");

    let grids = [];
    for (let g = 0; g < 50; g++) {
        let gridStr = "";
        for (let l = 1; l < 10; l++) {
            gridStr += arr[10 * g + l];
        }
        grids.push(gridStr.split("").map(x => Number(x)));
    }

    return grids;
}

/*  
    We want the grids to be global variables.

    A grid i an array with indices 0-80.
    A value of 0 means a blank cell.
    A grid has rows 1-9 and columns 1-9 and 9 boxes
    arranged as

    1 2 3
    4 5 6
    7 8 9
*/
let grids = syncReadFile('./euler_096.txt');

/*
    From index to (row, col) and back
*/
const i2rc = (i) => {
    return [Math.floor(i / 9) + 1, (i % 9) + 1];
}

const rc2i = (r, c) => {
    return 9 * (r - 1) + (c - 1);
}

/*
    Checking the validity of rows, columns and boxes
*/
const isValid = (arr) => {
    // Must contain the numbers 1-9
    return arr.sort((a, b) => a - b).join("") === "123456789";
}

const getRow = (grid, rowNo) => {
    const row = [...Array(81).keys()].filter(x => i2rc(x)[0] === rowNo).map(x => grid[x]);
    return row;
}

const getCol = (grid, colNo) => {
    const col = [...Array(81).keys()].filter(x => i2rc(x)[1] === colNo).map(x => grid[x]);
    return col;
}

const getBox = (grid, boxNo) => {
    let rowOffset = 0;
    let colOffset = 0;
    switch (boxNo) {
        case 1:
            break;
        case 2:
            colOffset = 3;
            break;
        case 3:
            colOffset = 6;
            break;
        case 4:
            rowOffset = 3;
            break;
        case 5:
            rowOffset = 3;
            colOffset = 3;
            break;
        case 6:
            rowOffset = 3;
            colOffset = 6;
            break;
        case 7:
            rowOffset = 6;
            break;
        case 8:
            rowOffset = 6;
            colOffset = 3;
            break;
        case 9:
            rowOffset = 6;
            colOffset = 6;
            break;
    }
    let box = [];
    for (let r = 1; r < 4; r++) {
        for (let c = 1; c < 4; c++) {
            box.push(grid[rc2i(r + rowOffset, c + colOffset)]);
        }
    }
    return box;
}

/*
    Finding blank cells and possible values of cells
*/
const blankCells = (grid) => {
    return [...Array(81).keys()].filter(x => grid[x] === 0);
}

const isPossible = (grid, i, val) => {
    const rowNo = i2rc(i)[0];
    const colNo = i2rc(i)[1];
    const boxNo = 3 * Math.floor((rowNo - 1) / 3) + Math.floor((colNo - 1) / 3) + 1;
    if (getRow(grid, rowNo).includes(val) || getCol(grid, colNo).includes(val) || getBox(grid, boxNo).includes(val)) {
        return false;
    }
    return true;
}


const possibleVals = (grid, i) => {
    return [1, 2, 3, 4, 5, 6, 7, 8, 9].filter(x => isPossible(grid, i, x));
}


/*
    Solving the sudoku with backtracking
*/
const solve = (gridNo, index) => {
    while (index < 81 && grids[gridNo][index]) {
        index++;
    }

    if (index === 81) {return true;}

    const pvs = possibleVals(grids[gridNo], index);

    for (const pv of pvs) {
        grids[gridNo][index] = pv;
        if (solve(gridNo, index + 1)) {
            return true;
        }
    }

    grids[gridNo][index] = 0;

    return false;
}

/*
    Collecting and adding
*/
let sum = 0;

for (let i = 0; i < 50; i++) {
    solve(i, 0);
    sum += Number([grids[i][0], grids[i][1], grids[i][2]].join(""));
}


const end = Date.now();

console.log("Execute time: ", end - start, " Sum of top left corners is: ", sum)